src/components/tl/self/tl_self_coll.c

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021-2022.  ALL RIGHTS RESERVED.
+ * Copyright (c) Meta Platforms, Inc. and affiliates. 2022.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "components/mc/base/ucc_mc_base.h"
+#include "components/mc/ucc_mc.h"
+#include "tl_self.h"
+#include "utils/ucc_coll_utils.h"
+#include "utils/ucc_malloc.h"
+
+static inline ucc_tl_self_task_t *
+ucc_tl_self_coll_init_task(ucc_base_coll_args_t *coll_args,
+                           ucc_base_team_t      *team)
+{
+    ucc_tl_self_team_t    *tl_team = ucc_derived_of(team, ucc_tl_self_team_t);
+    ucc_tl_self_context_t *ctx     = UCC_TL_SELF_TEAM_CTX(tl_team);
+    ucc_tl_self_task_t    *task    = ucc_mpool_get(&ctx->req_mp);
+
+    ucc_coll_task_init(&task->super, coll_args, team);
+    UCC_TL_SELF_PROFILE_REQUEST_NEW(task, "tl_self_task", 0);
+    task->super.finalize       = ucc_tl_self_coll_finalize;
+    task->super.triggered_post = ucc_triggered_post;
+    task->src = task->dst = NULL;
+    task->size            = 0;
+    return task;
+}
+
+static inline void ucc_tl_self_put_task(ucc_tl_self_task_t *task)
+{
+    UCC_TL_SELF_PROFILE_REQUEST_FREE(task);
+    ucc_mpool_put(task);
+}
+
+ucc_status_t ucc_tl_self_coll_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_self_task_t *task = ucc_derived_of(coll_task, ucc_tl_self_task_t);
+
+    tl_trace(UCC_TASK_LIB(task), "finalizing task %p", task);
+    ucc_tl_self_put_task(task);
+    return UCC_OK;
+}
+
+void ucc_tl_self_noop_progress(ucc_coll_task_t *task)
+{
+    task->status = UCC_OK;
+}
+
+ucc_status_t ucc_tl_self_noop_start(ucc_coll_task_t *task)
+{
+    task->progress = ucc_tl_self_noop_progress;
+    return ucc_progress_queue_enqueue(UCC_TASK_CORE_CTX(task)->pq, task);
+}
+
+void ucc_tl_self_coll_copy_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_self_task_t *task = ucc_derived_of(coll_task, ucc_tl_self_task_t);
+
+    task->super.status = ucc_mc_memcpy(task->dst, task->src, task->size,
+                                       task->dst_memtype, task->src_memtype);
+}
+
+ucc_status_t ucc_tl_self_coll_start(ucc_coll_task_t *task)
+{
+    return ucc_progress_queue_enqueue(UCC_TASK_CORE_CTX(task)->pq, task);
+}
+
+ucc_status_t ucc_tl_self_coll_noop_init(ucc_tl_self_task_t *task)
+{
+    task->super.status   = UCC_OK;
```

Comment:
status is set during ucc_coll_task_init to OPERATION_INITIALIZED. You should not reset it to UCC_OK.